web_services/foxx/brightid5/index.js

```diff
@@ -0,0 +1,324 @@
+'use strict';
+const secp256k1 = require('secp256k1');
+const createKeccakHash = require('keccak');
+
+const createRouter = require('@arangodb/foxx/router');
+const joi = require('joi');
+const arango = require('@arangodb').db;
+const nacl = require('tweetnacl');
+const db = require('./db');
+const schemas = require('./schemas').schemas;
+const operations = require('./operations');
+const initdb = require('./initdb').initdb;
+const {
+  strToUint8Array,
+  b64ToUint8Array,
+  uInt8ArrayToB64,
+  hash,
+  pad32,
+  addressToBytes32
+} = require('./encoding');
+
+initdb();
+const router = createRouter();
+module.context.use(router);
+const operationsHashesColl = arango._collection('operationsHashes');
+
+const TIME_FUDGE = 60 * 60 * 1000; // timestamp can be this far in the future (milliseconds) to accommodate client/server clock differences
+// error numbers
+const CONTEXT_NOT_FOUND = 1;
+const CONTEXTID_NOT_FOUND = 2;
+const CAN_NOT_BE_VERIFIED = 3;
+const NOT_SPONSORED = 4;
+const OLD_ACCOUNT = 5;
+const KEYPAIR_NOT_SET = 6;
+const ETHPRIVATEKEY_NOT_SET = 7;
+const ETHNAME_NOT_SET = 8;
+const OPERATION_NOT_FOUND = 9;
+const USER_NOT_FOUND = 10;
+const IP_NOT_SET = 11;
+const APP_NOT_FOUND = 12;
+
+const handlers = {
+  operationsPost: function(req, res){
+    const op = req.body;
+    op._key = hash(operations.getMessage(op));
+    if (operationsHashesColl.exists(op._key)) {
+      res.throw(400, 'operation is applied before');
+    }
+    try {
+      operations.verify(op);
+      // allow 60 operations in 15 minutes window by default
+      const timeWindow = (module.context.configuration.operationsTimeWindow || 15 * 60) * 1000;
+      const limit = module.context.configuration.operationsLimit || 60;
+      operations.checkLimits(op, timeWindow, limit);
+      if (op.name == 'Link ContextId') {
+        operations.encrypt(op);
+      }
+      else if (op.name == 'Sponsor') {
+        operations.updateSponsorOp(op);
+      }
+      op.state = 'init';
+      db.upsertOperation(op);
+    } catch (e) {
+      const code = (e == 'Too Many Requests') ? 429 : 400;
+      res.throw(code, e);
+    }
+    res.send({
+      data: {
+        _key: op._key
```

Comment:
Shouldn't this be named `hash` to be consistent with the `GET /operations/:hash` endpoint?  `id` is the most normal name for a parameter that is used to reference new object created by a `POST`. `hash` is unusual, but if the user should be aware it's a hash because they can use a hash function to do some checking on their side, I think it's fine to call it that.  `_key` is unusual, and only makes sense in the context of Arangodb, which I don't think the client needs to know about.

Reply:
Can't agree anymore!
The only place that should be changed is not here to use `hash` instead of `_key` for operations.
This was challenging so I ignored that, but as we have a breaking upgrade now for `v5` now for operations and we started documenting for app integration, I think you're right and we should apply this change just right now.
I pushed required changes here:
https://github.com/BrightID/BrightID-Node/commit/339a59fb852120fe5ebc702f3aba770dccc88b21
Hope not missed anything.